warp/Network/Wai/Handler/Warp/Request.hs

```diff
@@ -251,26 +256,35 @@ push maxTotalHeaderLength src (THStatus len lines prepend) bs'
         when (S.null bst) $ throwIO IncompleteHeaders
         push maxTotalHeaderLength src status bst
       where
-        len' = len + bsLen
         prepend' = S.append bs
-        status = THStatus len' lines prepend'
+        thisChunkLen = S.length bs'
+        newChunkLen = chunkLen + thisChunkLen
+        status = THStatus totalLen newChunkLen lines prepend'
     -- Found a newline, but next line continues as a multiline header
     push' (Just (end, True)) = push maxTotalHeaderLength src status rest
       where
         rest = S.drop (end + 1) bs
         prepend' = S.append (SU.unsafeTake (checkCR bs end) bs)
-        len' = len + end
-        status = THStatus len' lines prepend'
+        -- This is safe or we wouldn't have an 'end'
+        Just thisChunkUpToNL = findNewLine bs'
```

Comment:
I'd still rather see the partial match avoided. I think moving these clauses into `let`s within the `do` block would achieve that, plus avoid the unnecessary double call to `elemIndex`

Reply:
Made a new commit